Wikipedia/View Controllers/WebView/NativeSectionTitleOverlays/WMFSectionTitlesViewController.m

```diff
@@ -0,0 +1,231 @@
+    }
+    return self;
+}
+
+- (void)hideTopOverlay {
+    self.topStaticNativeTitleLabel.alpha = 0;
+}
+
+- (void)addOverlays {
+    for (WMFTitleOverlayModel* m in self.models) {
+        [m.label removeFromSuperview];
+    }
+    [self.models removeAllObjects];
+
+    [self setupTopStaticNativeTitleOverlayLabel];
+
+    UIView* browserView = [self.webView.scrollView wmf_firstSubviewOfClass:NSClassFromString(@"UIWebBrowserView")];
+
+    NSArray* sections = [self getSectionTitlesJSON];
+
+    for (NSDictionary* section in sections) {
+        NSNumber* sectionId = section[@"sectionId"];
+
+        NSString* title = section[@"text"];
+        if (title) {
+            title = [title wmf_stringByRemovingHTML];
+
+            WMFTitleOverlayLabel* label = [[WMFTitleOverlayLabel alloc] init];
+            label.text      = title;
+            label.sectionId = sectionId;
+
+            [self.webView.scrollView addSubview:label];
+
+            NSLayoutConstraint* (^ constrainEqually)(NSLayoutAttribute) = ^NSLayoutConstraint*(NSLayoutAttribute attr) {
+                NSLayoutConstraint* c =
+                    [NSLayoutConstraint constraintWithItem:label
+                                                 attribute:attr
+                                                 relatedBy:NSLayoutRelationEqual
+                                                    toItem:browserView
+                                                 attribute:attr
+                                                multiplier:1.0
+                                                  constant:0];
+
+                [self.webView.scrollView addConstraint:c];
+
+                return c;
+            };
+
+            constrainEqually(NSLayoutAttributeTrailing);
+            constrainEqually(NSLayoutAttributeLeading);
+
+            NSLayoutConstraint* topConstraint = constrainEqually(NSLayoutAttributeTop);
+
+            WMFTitleOverlayModel* m = [[WMFTitleOverlayModel alloc] init];
+            m.anchor        = section[@"anchor"];
+            m.title         = title;
+            m.topConstraint = topConstraint;
+            m.yOffset       = topConstraint.constant;
+            m.label         = label;
+            m.sectionId     = sectionId;
+            [self.models addObject:m];
+        }
+    }
+
+    [self updateOverlaysPositions];
+}
+
+- (void)setupTopStaticNativeTitleOverlayLabel {
+    if (self.topStaticNativeTitleLabel) {
+        return;
+    }
+    self.topStaticNativeTitleLabel       = [[WMFTitleOverlayLabel alloc] init];
+    self.topStaticNativeTitleLabel.alpha = 0;
+    [self.webViewController.view addSubview:self.topStaticNativeTitleLabel];
+    [self.topStaticNativeTitleLabel mas_makeConstraints:^(MASConstraintMaker* make) {
+        make.left.equalTo(self.webViewController.view.mas_left);
+        make.right.equalTo(self.webViewController.view.mas_right);
+    }];
+
+    self.topStaticNativeTitleLabelTopConstraint =
```

Comment:
I was able to switch the code to use mas_updateConstraints to vary the top constant of topStaticNativeTitleLabel as scrolling happens (i.e. nudging it when a section label needs to "push" it up offscreen). However, mas_updateConstraints  allocs a new MASConstraintMaker each time it runs, which itself allocs a new NSArray. We can't do this though because this code needs to run *each* time the scroll view scrolls so we need it to be as fast as possible, and simply maintaining a handle on "topStaticNativeTitleLabelTopConstraint" means we don't do *any* unnecessary alloc'ing.